tensorflow/core/platform/default/logging.cc

```diff
@@ -392,5 +345,182 @@ bool LogEveryNSecState::ShouldLog(double seconds) {
   return true;
 }
 
+std::mutex TFLogSinks::instanceMutex_;
+std::atomic<TFLogSinks*> TFLogSinks::instance_ = ATOMIC_VAR_INIT(nullptr);
+
+TFLogSinks& TFLogSinks::Instance() {
+  auto instance = instance_.load(std::memory_order_acquire);
+  if(!instance) {
+    std::lock_guard<std::mutex> lock(instanceMutex_);
+    instance = instance_.load(std::memory_order_relaxed);
+    if(!instance) {
+      instance = new TFLogSinks;
+      instance_.store(instance, std::memory_order_release);
+    }
+  }
+
+  return *instance;
+}
+
+void TFLogSinks::Add(TFLogSink* sink) {
+  if(sink == nullptr) {
+    return;
+  }
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto it = FindSink(sink);
+  if(it == sinks_.end()) {
+    sinks_.emplace_back(sink);
+  }
+}
+
+void TFLogSinks::Remove(TFLogSink* sink) {
+  if(sink == nullptr) {
+    return;
+  }
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto it = FindSink(sink);
+  if(it != sinks_.end()) {
+    sinks_.erase(it);
+  }
+}
+
+void TFLogSinks::Send(const TFLogEntry& entry) {
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Use `tensorflow::tf_shared_lock` for read access and `tensorflow::mutex_lock` for write access.

Reply:
Will do.

Reply:
There's no true "read" access in logging: there are potential cases of mutation, so just `tensorflow::mutex_lock`, unless y'all have an `upgrade_lock`.